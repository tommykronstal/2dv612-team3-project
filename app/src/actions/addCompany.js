import { UPDATE_COMPANY_NAME } from './types';

export function updateCompanyName (companyName) {
  return {
      type: UPDATE_COMPANY_NAME,
      companyName: companyName
  };
}

