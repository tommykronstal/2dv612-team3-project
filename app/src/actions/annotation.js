import {UPDATE_ANNOTATION} from './types'

export const updateAnnotation = (annotationText, materialId) => ({
  type: UPDATE_ANNOTATION,
  annotationText,
  materialId
})
