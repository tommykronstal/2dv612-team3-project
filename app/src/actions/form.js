import {CLEAR_FORM, SUBMIT_FORM, UPDATE_FIELD} from './types'

export const submitForm = (form, endpoint, action, tokenRequire = true, role = '') => ({
  type: SUBMIT_FORM,
  form,
  endpoint,
  action,
  tokenRequire,
  role
})

export const clearForm = form => ({
  type: CLEAR_FORM,
  form
})

export const updateField = (form, field, value) => ({
  type: UPDATE_FIELD,
  form,
  field,
  value
})
