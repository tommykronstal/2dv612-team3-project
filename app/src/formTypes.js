export const LOGIN = 'LOGIN'
export const ADD_COMPANY = 'ADD_COMPANY'
export const ADD_COMPANY_REP = 'ADD_COMPANY_REP'
export const REGISTER_CONSUMER = 'REGISTER_CONSUMER'
export const ADD_PRODUCT = 'ADD_PRODUCT'
export const UPDATE_PRODUCT = 'UPDATE_PRODUCT'
export const ADD_CATEGORY = 'ADD_CATEGORY'