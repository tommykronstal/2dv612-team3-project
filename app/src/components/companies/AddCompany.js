import React, { Component } from "react";
import {connect} from 'react-redux'

import {updateCompanyName} from '../../actions/addCompany'

import { action } from "@storybook/addon-actions";


import Button from "../Button";
import Section from "../Section";
import Loading from "../Loading";
import Header from "../Header";
import CenteredForm from "../CenteredForm";
import Content from "../Content";
import Title from "../Title";
import Input from "../Input";
import StatusModal from "../StatusModal";


function mapStateToProps({addCompany}) {
  return addCompany
}

function mapDispatchToProps(dispatch) {
  return {
    updateCompanyName: companyName => dispatch(updateCompanyName(companyName)),
  }
}

class AddCompany extends Component {
  render() {
    const {companyName} = this.props
    return (
      
      <div>
        <Content>
          <Title>Add a New Company</Title>
          <Input
            value={companyName}
            name="company"
            label="Company name"
            onChange={e => updateCompanyName(e.target.value)}
          />
          <Input
            type="email"
            name="email"
            label="Administrators E-Mail"
            onChange={action("changed")}
          />
          <Input
            type="password"
            name="password"
            label="Password"
            onChange={action("changed")}
          />
          <Button primary onClick={action("clicked")}>
            Add
          </Button>
        </Content>
      </div>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AddCompany);

// handleSubmit(e) {
//   this.setState(
//     {
//       newCompany: {
//         name: this.refs.name.value,
//         CompanyAdmin: {
//           name: this.refs.adminName.value,
//           email: this.refs.adminEmail.value
//         },
//         id: this.refs.name.value
//       }
//     },
//     function() {
//       //console.log(this.state.newCompany);
//       this.props.addCompany(this.state.newCompany);
//     }
//   );
//   e.preventDefault();
// }