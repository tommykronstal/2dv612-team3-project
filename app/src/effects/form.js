import {put, takeEvery, select, call} from 'redux-saga/effects'
import {SUBMIT_FORM, TOGGLE_LOADING, SET_STATUS} from '../actions/types'
import {clearForm} from '../actions/form'
import {post} from '../lib/http'
import {UPDATE_PRODUCT} from '../formTypes'

export function* watchFormActions() {
  yield takeEvery(SUBMIT_FORM, formRequest)
}

export function* formRequest({endpoint, form, action, tokenRequired}) {
  yield put({type: TOGGLE_LOADING})


  const {token, payload} = yield select(state => ({
    token: state.auth.jwt,
    payload: {
      ...state.form[form],
    },
  }))

  const tokenHeader = {...(token && tokenRequired && {Authorization: token})}

  if (form === UPDATE_PRODUCT) {
    yield sendProductFile(endpoint, token, payload, tokenHeader, action, form)
    return
  }

  console.log(payload)

  const response = yield call(post, endpoint, {
    headers: {
      ...tokenHeader,
    },
    body: JSON.stringify(payload)
  })

  console.log('response', response)

  yield handleResponse(response, form, action)
}

function* handleResponse(response, form, action) {
  if (response.status === 200 || response.status === 201) {
    if (action) {
      yield put(action(response))
    }
    yield put(clearForm(form))
  } else {
    if (response.message)
      yield put({type: SET_STATUS, warning: true, message: response.message})
  }

  yield put({type: TOGGLE_LOADING})
}


/**
 * Used to send data as a form.
 * Need to send it like this since original function to submit forms
 * assumes that it's JSON. (Maybe needs some tweaking in the future...)
 */
function* sendProductFile(endpoint, token, payload, tokenHeader = {}, action, form) {
  const formData = new FormData()

  console.log(payload)

  for( let key in payload) {
    formData.append(key, payload[key])
  }

  const response = yield call(post, endpoint, {
    headers: {
      ...tokenHeader
    },
    isJsonPayload: false,
    body: formData
  })

  yield handleResponse(response, form, action)
}