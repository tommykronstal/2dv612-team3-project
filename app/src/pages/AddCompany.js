import React from 'react'
import {connect} from 'react-redux'

import {ADD_COMPANY} from '../formTypes'
import {updateField, submitForm} from '../actions/form'
import { setStatus } from '../actions/status'
import Content from '../components/common/Content'
import Title from '../components/common/Title'
import Input from '../components/common/Input'
import Button from '../components/common/Button'

const AddCompany = props => (
  <Content>
    <form onSubmit={event => { 
      event.preventDefault() 
      props.addCompany()
    }}>
      <Title>Add a New Company</Title>
      <Input
        value={props.form.companyName}
        type='text'
        name='companyName'
        label='Company name'
        onChange={props.updateField}
      />
      <Input
        value={props.form.firstName}
        type='text'
        name='firstName'
        label='Administrators First Name'
        onChange={props.updateField}
      />
      <Input
        value={props.form.lastName}
        type='text'
        name='lastName'
        label='Administrators Last Name'
        onChange={props.updateField}
      />
      <Input
        value={props.form.email}
        type='email'
        name='email'
        label='Administrators E-Mail'
        onChange={props.updateField}
      />
      <Input
        value={props.form.password}
        type='password'
        name='password'
        label='Password'
        onChange={props.updateField}
      />
      <Button primary loading={props.loading.isLoading}>
        Add
      </Button>
    </form>
  </Content>
)

export default connect(
  ({loading, form}) => ({
    loading, 
    form: form[ADD_COMPANY] || {}
  }),
  dispatch => ({
    addCompany: () => dispatch(submitForm(
      ADD_COMPANY, 
      '/api/company/register', 
      _ => setStatus('Company Created')
    )),
    updateField: ({target}) => dispatch(updateField(ADD_COMPANY, target.name, target.value))   
  })
)(AddCompany)
