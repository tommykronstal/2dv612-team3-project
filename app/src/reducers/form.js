import { CLEAR_FORM, UPDATE_FIELD } from '../actions/types'
import { REGISTER_CONSUMER } from '../formTypes'
import USER_TYPES from '../userTypes'

const initialState = {
  [REGISTER_CONSUMER]: {
    role: USER_TYPES.USER
  }
}

export default function status (state = initialState, action = {}) {
  switch (action.type) {
  case CLEAR_FORM:
    return {
      ...state,
      [action.form]: initialState[action.form] || {}
    }
  case UPDATE_FIELD:
    return {
      ...state,
      [action.form]: {
        ...state[action.form],
        [action.field]: action.value
      }
    }
  default:
    return state
  }
}
