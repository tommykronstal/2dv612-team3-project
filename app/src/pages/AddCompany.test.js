import React from 'react'
import renderer from 'react-test-renderer'

import { AddCompany, mapStateToProps, mapDispatchToProps, handleResponse } from './AddCompany'
import { ADD_COMPANY } from '../formTypes'
import { submitForm, updateField } from '../actions/form'

describe('AddCompany', () => {

  describe('component', () => {
    it('renders correctly, initial', () => {
      const props = {
        addCompany: jest.fn(),
        updateField: jest.fn(),
        form: {},
        loading: false
      }
  
      const tree = renderer.create(<AddCompany {...props} />).toJSON()
  
      expect(tree).toMatchSnapshot()
    })
  
    it('renders correctly, with data', () => {
      const props = {
        addCompany: jest.fn(),
        updateField: jest.fn(),
        form: {
          companyName: 'company!', 
          firstName: 'kalle',
          lastName: 'kula',
          email: 'woo@woo.com', 
          password: 'aaa' 
        },
        loading: true
      }
  
      const tree = renderer.create(<AddCompany {...props} />).toJSON()
  
      expect(tree).toMatchSnapshot()
    })
  })

  describe('state', () => {
    it('should have access to representative form and loading state', () => {
      const state = {
        loading: { isLoading: false },
        form: { [ADD_COMPANY]: {username: 'woooo'}}
      }
    
      expect(mapStateToProps(state)).toEqual({ 
        loading: state.loading, 
        form: state.form[ADD_COMPANY] 
      })
    })
  })

  describe('actions', () => {
    it('should be able to dispatch submit form actions', () => {
      const dispatch = jest.fn()
    
      const props = mapDispatchToProps(dispatch)
      props.addCompany()
    
      expect(dispatch).toBeCalledWith(submitForm(
        ADD_COMPANY, 
        '/api/company', 
        handleResponse
      ))
    })

    it('should be able to dispatch update field actions', () => {
      const dispatch = jest.fn()
      const event = {
        target: {
          name: 'user',
          value: 'kalle'
        }
      }
    
      const props = mapDispatchToProps(dispatch)
      props.updateField(event)
    
      expect(dispatch).toBeCalledWith(updateField(ADD_COMPANY, event.target.name, event.target.value))
    })
  })
})
