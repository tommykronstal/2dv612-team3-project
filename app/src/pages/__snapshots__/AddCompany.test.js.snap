// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCompany component renders correctly, initial 1`] = `
<section
  className="sc-bdVaJa cgbLUC"
>
  <form
    className="sc-bwzfXH gJIwTT"
    onSubmit={[Function]}
  >
    <h1
      className="sc-htpNat cfPrNC"
    >
      Add a New Company
    </h1>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="companyName"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        className="sc-ifAKCX crRGrj"
        value=""
      >
        Company name
      </label>
    </div>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="firstName"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        className="sc-ifAKCX crRGrj"
        value=""
      >
        Administrators First Name
      </label>
    </div>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="lastName"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        className="sc-ifAKCX crRGrj"
        value=""
      >
        Administrators Last Name
      </label>
    </div>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="email"
        onChange={[Function]}
        type="email"
        value=""
      />
      <label
        className="sc-ifAKCX crRGrj"
        value=""
      >
        Administrators E-Mail
      </label>
    </div>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
      <label
        className="sc-ifAKCX crRGrj"
        value=""
      >
        Password
      </label>
    </div>
    <button
      className="sc-gzVnrw jmsTpF"
      disabled={undefined}
      onClick={[Function]}
    >
      Add
    </button>
  </form>
</section>
`;

exports[`AddCompany component renders correctly, with data 1`] = `
<section
  className="sc-bdVaJa cgbLUC"
>
  <form
    className="sc-bwzfXH gJIwTT"
    onSubmit={[Function]}
  >
    <h1
      className="sc-htpNat cfPrNC"
    >
      Add a New Company
    </h1>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="companyName"
        onChange={[Function]}
        type="text"
        value="company!"
      />
      <label
        className="sc-ifAKCX bXngpg"
        value="company!"
      >
        Company name
      </label>
    </div>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="firstName"
        onChange={[Function]}
        type="text"
        value="kalle"
      />
      <label
        className="sc-ifAKCX bXngpg"
        value="kalle"
      >
        Administrators First Name
      </label>
    </div>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="lastName"
        onChange={[Function]}
        type="text"
        value="kula"
      />
      <label
        className="sc-ifAKCX bXngpg"
        value="kula"
      >
        Administrators Last Name
      </label>
    </div>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="email"
        onChange={[Function]}
        type="email"
        value="woo@woo.com"
      />
      <label
        className="sc-ifAKCX bXngpg"
        value="woo@woo.com"
      >
        Administrators E-Mail
      </label>
    </div>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="password"
        onChange={[Function]}
        type="password"
        value="aaa"
      />
      <label
        className="sc-ifAKCX bXngpg"
        value="aaa"
      >
        Password
      </label>
    </div>
    <button
      className="sc-gzVnrw jmsTpF"
      disabled={undefined}
      onClick={[Function]}
    >
      Add
    </button>
  </form>
</section>
`;
